src/NServiceBus.Core/Routing/EndpointInstances.cs

```diff
@@ -2,106 +2,67 @@ namespace NServiceBus.Routing
 {
     using System;
     using System.Collections.Generic;
-    using System.Threading.Tasks;
+    using System.Linq;
 
     /// <summary>
     /// Stores the information about instances of known endpoints.
     /// </summary>
     public class EndpointInstances
     {
-        internal Task<IEnumerable<EndpointInstance>> FindInstances(string endpoint)
+        internal IEnumerable<EndpointInstance> FindInstances(string endpoint)
         {
-            HashSet<EndpointInstance> staticInstances;
-            staticRules.TryGetValue(endpoint, out staticInstances);
-
-            if (dynamicRules.Count > 0)
-            {
-                return FindDynamicInstances(endpoint, staticInstances);
-            }
-
-            if (staticInstances != null)
-            {
-                return Task.FromResult<IEnumerable<EndpointInstance>>(staticInstances);
-            }
-
-            return Task.FromResult<IEnumerable<EndpointInstance>>(new[]
-            {
-                new EndpointInstance(endpoint)
-            });
+            HashSet<EndpointInstance> instances;
+            return cache.TryGetValue(endpoint, out instances)
+                ? instances
+                : EnumerableEx.Single(new EndpointInstance(endpoint));
```

Comment:
not a big fan of the `EnumerableEx` extensions. I think a new[] { new EndpointInstance(endpoint) } would be more readable here?